src/bgw/job_stat_history.c

```diff
@@ -21,29 +23,87 @@ typedef struct BgwJobStatHistoryContext
 	Jsonb *edata;
 } BgwJobStatHistoryContext;
 
+static Jsonb *
+build_job_info(BgwJob *job)
+{
+	JsonbParseState *parse_state = NULL;
+	pushJsonbValue(&parse_state, WJB_BEGIN_OBJECT, NULL);
+
+	/* all fields that is possible to change with `alter_job` API */
+	ts_jsonb_add_interval(parse_state, "schedule_interval", &job->fd.schedule_interval);
+	ts_jsonb_add_interval(parse_state, "max_runtime", &job->fd.max_runtime);
+	ts_jsonb_add_int32(parse_state, "max_retries", job->fd.max_retries);
+	ts_jsonb_add_interval(parse_state, "retry_period", &job->fd.retry_period);
+	ts_jsonb_add_str(parse_state, "proc_schema", NameStr(job->fd.proc_schema));
+	ts_jsonb_add_str(parse_state, "proc_name", NameStr(job->fd.proc_name));
+	ts_jsonb_add_str(parse_state, "owner", GetUserNameFromId(job->fd.owner, false));
+	ts_jsonb_add_bool(parse_state, "scheduled", job->fd.scheduled);
+	ts_jsonb_add_bool(parse_state, "fixed_schedule", job->fd.fixed_schedule);
+
+	if (job->fd.initial_start)
+		ts_jsonb_add_interval(parse_state, "initial_start", &job->fd.retry_period);
+
+	if (job->fd.hypertable_id != INVALID_HYPERTABLE_ID)
+		ts_jsonb_add_int32(parse_state, "hypertable_id", job->fd.hypertable_id);
+
+	if (strlen(NameStr(job->fd.check_schema)) > 0)
+		ts_jsonb_add_str(parse_state, "check_schema", NameStr(job->fd.check_schema));
+
+	if (strlen(NameStr(job->fd.check_name)) > 0)
+		ts_jsonb_add_str(parse_state, "check_name", NameStr(job->fd.check_name));
+
+	if (job->fd.timezone)
+		ts_jsonb_add_str(parse_state, "timezone", text_to_cstring(job->fd.timezone));
+
+	return JsonbValueToJsonb(pushJsonbValue(&parse_state, WJB_END_OBJECT, NULL));
+}
+
+static Jsonb *
+ts_bgw_job_stat_history_build_data_info(BgwJob *job, Jsonb *error_data)
+{
+	JsonbParseState *parse_state = NULL;
+	JsonbValue value = { 0 };
+	pushJsonbValue(&parse_state, WJB_BEGIN_OBJECT, NULL);
+
+	Assert(job != NULL);
+
+	/* job information jsonb */
+	JsonbToJsonbValue(build_job_info(job), &value);
+	ts_jsonb_add_value(parse_state, "job", &value);
+
+	if (job->fd.config != NULL)
+	{
+		/* config information jsonb*/
+		JsonbToJsonbValue(job->fd.config, &value);
+		ts_jsonb_add_value(parse_state, "config", &value);
+	}
+
+	if (error_data != NULL)
+	{
+		/* error information jsonb */
+		JsonbToJsonbValue(error_data, &value);
+		ts_jsonb_add_value(parse_state, "error_data", &value);
+	}
+
+	return JsonbValueToJsonb(pushJsonbValue(&parse_state, WJB_END_OBJECT, NULL));
+}
+
 static void
-ts_bgw_job_stat_history_insert(BgwJob *job, BgwJobStatHistoryContext *context)
+ts_bgw_job_stat_history_insert(BgwJobStatHistoryContext *context)
 {
 	Relation rel = table_open(catalog_get_table_id(ts_catalog_get(), BGW_JOB_STAT_HISTORY),
 							  ShareRowExclusiveLock);
 	TupleDesc desc = RelationGetDescr(rel);
 	NullableDatum values[Natts_bgw_job_stat_history] = { { 0 } };
 	CatalogSecurityContext sec_ctx;
 
-	ts_datum_set_int32(Anum_bgw_job_stat_history_job_id, values, job->fd.id, false);
```

Comment:
I am wondering if it is valid to call the function with `context = NULL`. If not, could you add an assert to the beginning of the function?

Reply:
Make sense... also did a small refactoring to pass only the context because it already contains the BgwJob